Getting started chapter (#1)
nanjekyejoannah/micropython: docs/develop/gettingstarted.rst

.. _gettingstarted:

Getting Started
===============

This guide covers a step by step process on setting up version control, obtaining and building

> a copy of the source code on a port, building the Docs, running tests and a description of the 

Reviewer: ```suggestion
a copy of the source code for a port, building the Docs, running tests and a description of the 
```

code is pulled and pushed to and from the main repository. Install the respective version of ``Git`` for 
your operating system to follow through the rest of the steps.

.. note::
   For a reference on the installation instructions, please refer to 
   the `Git installation instructions <https://git-scm.com/book/en/v2/Getting-Started-Installing-Git>`_.
   Learn about the basic git commands in this `Git Handbook <https://guides.github.com/introduction/git-handbook/>`_
   or any other sources on the internet.

Get the code
----------------

It is recommended that you maintain a fork of the MicroPython repository for your development purposes.
The process of obtaining the source code includes the following:

#. Fork the repository https://github.com/micropython/micropython.git
#. You will now have a fork at <https://github.com/<yourusername>/micropython.git>.
#. Clone the repository. You may clone either your fork or the main MicroPython repository.
   Use one of the commands below:

.. code-block:: console

   $ git clone https://github.com/micropython/micropython.git 
   $ git clone https://github.com/<yourusername>/micropython.git

> Depending on what you cloned, do the necessary configuration.

Reviewer: Users may need some clarity around what "necessary configuration" means.

Configure upstream if you cloned your fork.

.. code-block:: console

   $ cd micropython
   $ git remote add upstream https://github.com/micropython/micropython.git

> Configure upstream and origin if you cloned the main repository. You can maintain your

Reviewer: Maybe something like:
```suggestion
It is common to configure ``upstream`` and ``origin`` on a forked repository to assist with sharing code changes. You can maintain your
```

   upstream	https://github.com/micropython/micropython.git (push)

You should now have a copy of the source code. By default you are pointing
to the master branch. To get ready for further development, it is recommended
to work on a development branch.

.. code-block:: console

    $ git checkout -b dev-branch

You can give it any name. You will have to compile MicroPython whenever you change 
to a different branch.

Build the Code
--------------

For development, you should compile MicroPython for a specific platform which corresponds
to a port. Before any building for a port, first build the MicroPython cross-compiler.
Install the required dependencies for the port in order to sucessfully
compile and build.

Build the cross-compiler
~~~~~~~~~~~~~~~~~~~~~~~~

Almost all ports require building ``mpy-cross`` first to perform pre-compilation
of scripts that will be included in the port firmware:

.. code-block:: console

   $ cd mpy-cross
   $ make

If ``mpy-cross`` built successfully, you should see a message similar to this:

.. code-block:: console

   LINK mpy-cross
      text	   data	    bss	    dec	    hex	filename
    279328	    776	    880	 280984	  44998	mpy-cross

> The unix port

Reviewer: We may need to provide some guidance on what the unix port is. I imagine many beginners wouldn't know if they need to build it or not! 
The description of what the unix port is doesn't necessarily have to be _here_ though. A link would be fine as long as it will help a user understand if they need to build the port or not.

~~~~~~~~~~~~~
To build for the unix port, make sure you have a working environment for ``gcc`` and ``GNU make``.

> All Unixes come with ``gcc`` installed. Confirm if unsure:

Reviewer: As discussed: Not all unixes come with gcc _installed_, though it's usually easy to install.

.. code-block:: console

   $ gcc --version

>    gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609

Reviewer: That version number reminds me - you may want to explain at the top something like: Ubuntu 20.04 has been used for this example but other unixes ought to work with little modification.

   Copyright (C) 2015 Free Software Foundation, Inc.
   This is free software; see the source for copying conditions.  There is NO
   warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Install the required dependencies:

.. code-block:: console

   $ sudo apt-get install build-essential libreadline-dev libffi-dev git \\
   pkg-config gcc-arm-none-eabi libnewlib-arm-none-eabi

Then build:

.. code-block:: console

   $ cd ports/unix
   $ make submodules
   $ make

If MicroPython built correctly, you should see the following:

.. code-block:: console

   LINK micropython
      text	   data	    bss	    dec	    hex	filename
    412033	   5680	   2496	 420209	  66971	micropython

Now run it:

.. code-block:: console

   $ ./micropython
   MicroPython v1.13-38-gc67012d-dirty on 2020-09-13; linux version
   Use Ctrl-D to exit, Ctrl-E for paste mode
   >>> print("hello world")
   hello world
   >>>

The stm32 port
~~~~~~~~~~~~~~

Like the unix port, you need to install some required dependencies.

> The ARM compiler is required:

Reviewer: ```suggestion
The ARM cross-compiler is required:
```

.. code-block:: console

   $ sudo apt-get install arm-none-eabi-gcc arm-none-eabi-binutils arm-none-eabi-newlib

Build:

.. code-block:: console

   $ cd ports/stm32
   $ make submodules
   $ make

Please refer to the `stm32 documentation <https://github.com/micropython/micropython/tree/master/ports/stm32>`_ 
for more details on flashing the firmare.

> The windows port

Reviewer: Windows should probably be capitalized:

```suggestion
The Windows port
```
